extensions/amp-story/1.0/amp-story.js

```diff
@@ -2375,9 +2360,32 @@ export class AmpStory extends AMP.BaseElement {
       nextPageEl.setAttribute(Attributes.RETURN_TO, pageToBeInsertedId);
     }
 
+    // Adjust the ad page's position in this.adPages_ array to reflect the actual.
+    const adPageIndex = this.getPageIndexById(pageToBeInsertedId);
+    if (adPageIndex != -1) {
+      this.pages_.splice(adPageIndex, 1);
+    }
+    this.pages_.splice(
+      this.getPageIndexById(pageBeforeId) + 1,
+      0,
+      pageToBeInserted
+    );
+
+    this.insertPageToPageIds_(pageBeforeId, pageToBeInsertedId);
     return true;
   }
 
+  /** @private */
+  insertPageToPageIds_(pageBeforeId, pageToBeInsertedId) {
+    const pageIds = [...this.storeService_.get(StateProperty.PAGE_IDS)];
+    const insertPos = pageIds.indexOf(pageBeforeId);
+    if (insertPos === -1 || pageIds.includes(pageToBeInsertedId)) {
+      return;
+    }
+    pageIds.splice(insertPos + 1, 0, pageToBeInsertedId);
+    this.storeService_.dispatch(Action.SET_PAGE_IDS, pageIds);
+  }
```

Comment:
Re: you should really do `this.pages_.map((el) => el.id)`. The DOM is not ordered (I really believe it should be tho, but it's harder), but you just ordered the `this.pages_` array above.

Reply:
Done.